include/RAJA/policy/cuda/reduce.hpp

```diff
@@ -109,1507 +390,1185 @@ __device__ __forceinline__ T shfl_xor(T var, int laneMask)
-    }
-#endif
-  }
-
-  /*!
-   * \brief Operator that returns reduced sum value.
-   *
-   * Note: accessor only executes on host.
-   */
-  operator T()
-  {
-    beforeCudaReadTallyBlock<Async>(m_myID);
-    return m_tally_host->tally;
-  }
-
-  /*!
-   * \brief Method that returns reduced sum value.
-   *
-   * Note: accessor only executes on host.
-   */
-  T get() { return operator T(); }
-
-  /*!
-   * \brief Operator that adds value to sum.
-   *
-   * Note: only operates on device.
-   */
-  __device__ ReduceSum<cuda_reduce<BLOCK_SIZE, Async>, T> const &operator+=(
-      T val) const
-  {
-    extern __shared__ unsigned char sd_block[];
-    T *sd = reinterpret_cast<T *>(&sd_block[m_smem_offset]);
-
-    int threadId = threadIdx.x + blockDim.x * threadIdx.y
-                   + (blockDim.x * blockDim.y) * threadIdx.z;
 
-    sd[threadId] += val;
+      __syncthreads();
 
+    }
 
-    return *this;
+    return temp;
   }
+};
 
-private:
-  /*!
-   * \brief Default constructor is declared private and not implemented.
-   */
-  ReduceSum<cuda_reduce<BLOCK_SIZE, Async>, T>();
-
-  /*!
-   * \brief Pointer to host tally block cache slot for this reduction variable.
-   */
-  CudaReductionTallyType<T> *m_tally_host = nullptr;
-
-  /*!
-   * \brief Pointer to device data block for this reduction variable.
-   */
-  CudaReductionBlockType<T> *m_blockdata = nullptr;
-
-  /*!
-   * \brief Pointer to device tally block slot for this reduction variable.
-   */
-  CudaReductionTallyType<T> *m_tally_device = nullptr;
-
-  /*!
-   * \brief My cuda reduction variable ID.
-   */
-  int m_myID = -1;
 
-  /*!
-   * \brief Byte offset into dynamic shared memory.
-   */
-  int m_smem_offset = -1;
-
-  /*!
-   * \brief If this variable is a copy or not; only original may release memory
-   *        or perform finalization.
-   */
-  bool m_is_copy_host = false;
-  bool m_is_copy_device = false;
-  bool m_finish_reduction = false;
-
-  // Sanity checks for block size and template type size
-  static constexpr bool powerOfTwoCheck = (!(BLOCK_SIZE & (BLOCK_SIZE - 1)));
-  static constexpr bool reasonableRangeCheck =
-      ((BLOCK_SIZE >= 32) && (BLOCK_SIZE <= 1024));
-  static constexpr bool sizeofcheck =
-      ((sizeof(T) <= sizeof(CudaReductionDummyDataType))
-       && (sizeof(CudaReductionTallyType<T>)
-           <= sizeof(CudaReductionDummyTallyType))
-       && (sizeof(CudaReductionBlockType<T>)
-           <= sizeof(CudaReductionDummyBlockType)));
-  static_assert(powerOfTwoCheck, "Error: block sizes must be a power of 2");
-  static_assert(reasonableRangeCheck,
-                "Error: block sizes must be between 32 and 1024");
-  static_assert(sizeofcheck,
-                "Error: type must be of size <= " RAJA_STRINGIFY_MACRO(
-                    RAJA_CUDA_REDUCE_VAR_MAXSIZE));
-};
+//! Cuda Reduction Atomic entity -- generalize on reduction, and type
+template <bool Async, typename Reducer, typename T>
+struct CudaReduceAtomic {
```

Comment:
I'm confused. I thought we generalized on the Reducer type. Isn't Atomic dependent upon the reducer?